pkg/ebpf/events_enrich.go

```diff
@@ -8,172 +8,157 @@ import (
+				// CgroupMkdir: pick EventID from the event itself
 				if event.EventID == int(events.CgroupMkdir) {
 					cgroupId, _ = parse.ArgUint64Val(event, "cgroup_id")
 				}
-				// cgroup_rmdir: need to clean attempt so cgroupId can be reused
+				// CgroupRmdir: clean up remaining events and maps
 				if event.EventID == int(events.CgroupRmdir) {
 					cgroupId, _ = parse.ArgUint64Val(event, "cgroup_id")
-					attemptedMutex.Lock()
-					delete(attempted, cgroupId)
-					attemptedMutex.Unlock()
-				}
-				// non container event and not cgroup_mkdir (or the event already enriched) skip per cgroupId caching
-				if (event.ContainerID == "" && event.EventID != int(events.CgroupMkdir)) || event.ContainerImage != "" {
-					// enrichment attempt is being done, or has failed, so do not try it again
-					attemptedMutex.RLock()
-					if _, ok := attempted[cgroupId]; ok {
-						// wait transaction (cache flush) to finish so cached events and this one are ordered
-						attempted[cgroupId].Lock()
-						out <- event
-						attempted[cgroupId].Unlock()
-						attemptedMutex.RUnlock()
-						continue
-					}
-					attemptedMutex.RUnlock()
-					out <- event // some events might not have an attempted map
-					continue
-				} else {
-					//if the container queue has not been created - create the container queue and invoke to enrich query
-					queuesMutex.RLock()
-					_, exists := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if !exists {
-						queuesMutex.Lock()
-						queues[cgroupId] = make(chan *trace.Event, enrichmentQueueSize)
-						queuesMutex.Unlock()
-
-						go func() {
-							metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
-							enriches <- enrichResult{cgroupId, metadata, err}
-						}()
-					}
-					queuesMutex.RLock()
-					queue, ok := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if ok {
-						//if queue still exists send as usual
-						queue <- event
-					} else {
-						//if not, get the enriched data and handle the stale queue
-						info := t.containers.GetCgroupInfo(cgroupId)
-						queuesMutex.RLock()
-						queue = staleQueues[cgroupId]
-						queuesMutex.RUnlock()
-
-						//probably empty but just to make sure
-						for evt := range queue {
-							enrichEvent(evt, info.Container)
-							out <- evt
+					i := enrichInfo[cgroupId]
+					for evt := range queues[cgroupId] {
+						if evt.ContainerImage == "" {
+							enrichEvent(evt, i.result)
 						}
-
-						enrichEvent(event, info.Container)
-
-						close(queue)
-						queuesMutex.Lock()
-						delete(staleQueues, cgroupId)
-						queuesMutex.Unlock()
-						out <- event
+						out <- evt
 					}
+					bLock.Lock()
+					delete(enrichDone, cgroupId)
+					delete(enrichInfo, cgroupId)
+					delete(queues, cgroupId)
+					bLock.Unlock()
+					continue
 				}
+				// make sure a queue channel exists for this cgroupId
+				if _, ok := queues[cgroupId]; !ok {
+					queueSize := contQueueSize
+					bLock.Lock()
```

Comment:
I think you should take the lock before the check
`if _, ok := queues[cgroupId]; !ok {`
or otherwise, you might perform the above two lines more than once for the same cgroupId

Reply:
yep.